### PR TITLE
added the delete jobs and cancel job functionality 

### PR DESCRIPTION
**Pull request by Nike682631**
Signed-off-by: nikunj goyal <nikunjgoyal31@gmail.com>

### CODE REVIEW

File: src/main/java/zosjobs/CancelJobs.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     *
+     * @return job document with details about the submitted job
+     * @throws Exception error on submitting
+     * @author Frank Giordano
+     */
+    public Response cancelJob() throws Exception {
+        return this.cancelJobsCommon(null);
+    }
+
+
```
Comment (Nurkambay) [inline src/main/java/zosjobs/CancelJobs.java:40]: Extra line not needed here

File: src/main/java/zosjobs/CancelJobs.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    public Response cancelJobsCommon(CancelJobParams parms) throws Exception {
```
Comment (Nurkambay) [inline src/main/java/zosjobs/CancelJobs.java:41]: Please add a JavaDoc here

File: src/main/java/zosjobs/CancelJobs.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
+package zosjobs;
```
Comment (Nurkambay) [inline src/main/java/zosjobs/CancelJobs.java:1]: Please add a text header to this file

File: src/main/java/zosjobs/DeleteJobs.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
+package zosjobs;
```
Comment (Nurkambay) [inline src/main/java/zosjobs/DeleteJobs.java:1]: Please add a text header to this file

File: src/main/java/zosjobs/input/CancelJobParams.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Optional;
+
+public class CancelJobParams {
+    private Optional<String> jobname;
+    private Optional<String> jobId;
+    private Optional<String> version;
+
+    public CancelJobParams(CancelJobParams.Builder builder) {
+
+
```
Comment (Nurkambay) [inline src/main/java/zosjobs/input/CancelJobParams.java:12]: Extra line non needed here

File: src/main/java/zosjobs/input/DeleteJobParams.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
+package zosjobs.input;
```
Comment (Nurkambay) [inline src/main/java/zosjobs/input/DeleteJobParams.java:1]: Please add a text header to this file

File: src/main/java/zosjobs/input/DeleteJobParams.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.Optional;
+
+public class DeleteJobParams {
```
Comment (Nurkambay) [inline src/main/java/zosjobs/input/DeleteJobParams.java:5]: JavaDoc is needed for this class

File: src/main/java/zosjobs/input/DeleteJobParams.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+    private Optional<String> jobname;
+    private Optional<String> jobId;
+    private Optional<String> modifyVersion;
+
+    public DeleteJobParams(DeleteJobParams.Builder builder) {
+
+
```
Comment (Nurkambay) [inline src/main/java/zosjobs/input/DeleteJobParams.java:13]: Extra line is not needed here

File: src/main/java/zosjobs/input/DeleteJobParams.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        if (builder.jobname != null)
```
Comment (Nurkambay) [inline src/main/java/zosjobs/input/DeleteJobParams.java:14]: It is not necessary to do it.
You just can do the following:
```java
this.jobname = Optional.ofNullable(builder.jobname);
```

File: src/main/java/zosjobs/input/DeleteJobParams.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        if (builder.jobId != null)
+            this.jobId = Optional.ofNullable(builder.jobId);
+        else this.jobId = Optional.empty();
+
+        if (builder.modifyVersion != null)
+            this.modifyVersion = Optional.ofNullable(builder.modifyVersion);
+        else this.modifyVersion = Optional.empty();
+    }
+
+
```
Comment (Nurkambay) [inline src/main/java/zosjobs/input/DeleteJobParams.java:27]: Extra line

File: src/main/java/zosjobs/input/CancelJobParams.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        if (builder.jobname != null)
+            this.jobname = Optional.ofNullable(builder.jobname);
```
Comment (Nurkambay) [inline src/main/java/zosjobs/input/CancelJobParams.java:14]: It is not necessary to do it.
You just can do the following:
```java
this.jobname = Optional.ofNullable(builder.jobname);
```

**Review by Nurkambay — CHANGES_REQUESTED** [review]

File: src/main/java/zosjobs/CancelJobs.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Cancel a job that resides in a z/OS data set.
+     *
+     * @return job document with details about the submitted job
+     * @throws Exception error on submitting
+     * @author Frank Giordano
+     */
+    public Response cancelJob() throws Exception {
+        return this.cancelJobsCommon(null);
```
Comment (Nurkambay) [inline src/main/java/zosjobs/CancelJobs.java:37]: If params are null you will have an exception here:
```
Util.checkNullParameter(parms == null, "parms is null");
```
just create an empty params object here

File: src/main/java/zosjobs/CancelJobs.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+
+    public Response cancelJobsCommon(CancelJobParams parms) throws Exception {
+        Util.checkNullParameter(parms == null, "parms is null");
+        Util.checkStateParameter(!parms.getJobId().isPresent(), "jobid not specified");
+        Util.checkStateParameter(!parms.getJobname().isPresent(), "jobname not specified");
+
+        String url = "https://" + connection.getHost() + ":" + connection.getPort() + JobsConstants.RESOURCE + JobsConstants.REQUEST_CANCEL;
```
Comment (Nurkambay) [inline src/main/java/zosjobs/CancelJobs.java:46]: Why you do not have "/" here between JobsConstants.RESOURCE and JobsConstants.REQUEST_CANCEL ?

File: src/main/java/zosjobs/CancelJobs.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        LOG.debug(url);
+
+        Map<String, String> headers = new HashMap<>();
+        String key, value, version;
+
+        if(parms.getVersion().get() == null) {
```
Comment (Nurkambay) [inline src/main/java/zosjobs/CancelJobs.java:52]: use
```java
if (!parms.getVersion().isPresent())
```
instead

File: src/main/java/zosjobs/DeleteJobs.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+     *
+     * @return job document with details about the submitted job
+     * @throws Exception error on submitting
+     * @author Frank Giordano
+     */
+    public Response deletejob( ) throws Exception {
+        return this.deleteJobCommon(null);
+    }
+
+    public Response deleteJobCommon(DeleteJobParams parms) throws Exception {
```
Comment (Nurkambay) [inline src/main/java/zosjobs/DeleteJobs.java:40]: JavaDocs

File: src/main/java/zosjobs/CancelJobs.java (hunk 3 above)
Comment (frankgiordano) [reply src/main/java/zosjobs/CancelJobs.java:1]: license header.

File: src/main/java/zosjobs/CancelJobs.java (hunk 14 above)
Comment (frankgiordano) [reply src/main/java/zosjobs/CancelJobs.java:52]: the project is setup for java 11. with java 11 you can use the cleaner version of !parms.getVersion().isPresent() which is parms.getVersion().isEmpty() instead

File: src/main/java/zosjobs/CancelJobs.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    public CancelJobs(ZOSConnection connection) {
+        this.connection = connection;
+    }
+
+    /**
+     * Cancel a job that resides in a z/OS data set.
+     *
+     * @return job document with details about the submitted job
+     * @throws Exception error on submitting
+     * @author Frank Giordano
```
Comment (frankgiordano) [inline src/main/java/zosjobs/CancelJobs.java:34]: remove my name as @author and use yours instead for every instance for your code!

File: src/main/java/zosjobs/CancelJobs.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    public Response cancelJob() throws Exception {
+        return this.cancelJobsCommon(null);
+    }
+
+
+    public Response cancelJobsCommon(CancelJobParams parms) throws Exception {
+        Util.checkNullParameter(parms == null, "parms is null");
+        Util.checkStateParameter(!parms.getJobId().isPresent(), "jobid not specified");
+        Util.checkStateParameter(!parms.getJobname().isPresent(), "jobname not specified");
```
Comment (frankgiordano) [inline src/main/java/zosjobs/CancelJobs.java:44]: the project is setup for java 11. with java 11 you can use the cleaner version of !parms.getVersion().isPresent() which is parms.getJobname().isEmpty() instead etc..

File: src/main/java/zosjobs/DeleteJobs.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        Util.checkNullParameter(parms == null, "parms is null");
+        Util.checkStateParameter(!parms.getJobId().isPresent(), "jobid not specified");
+        Util.checkStateParameter(!parms.getJobname().isPresent(), "jobname not specified");
+
```
Comment (frankgiordano) [inline src/main/java/zosjobs/DeleteJobs.java:44]: it seems modifyversion is required as you are using it below and as such we such check for it before using it with Util.checkStateParameter(parms.getModifyVersion().isEmpty(), "modify version not specified"

File: src/main/java/zosjobs/input/DeleteJobParams.java [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public String toString() {
+        return "DeleteJobParms{" +
+                "jobId=" + jobId +
+                ", jobname=" + jobname +
+                ", modifyVersion=" + modifyVersion +
+                '}';
+    }
+
+    public static class Builder {
```
Comment (frankgiordano) [inline src/main/java/zosjobs/input/DeleteJobParams.java:47]: do we need to use builder pattern for this class that only has three members might be an overkill. seems all three members are required so not sure if optional builders are needed.. what do you think? use this without builder but with Optional.. are all three members are required?

**Review by frankgiordano — COMMENTED** [review]
see inline comments

File: src/main/java/zosjobs/input/DeleteJobParams.java (hunk 19 above)
Comment (Nike682631) [reply src/main/java/zosjobs/input/DeleteJobParams.java:59]: Hey Frank, I'm not clear about what I need to do here. I've resolved other comments and was hoping you could specify a bit more on what I should do in this? I'm currently marking this as resolved but feel free to unresolve it with a comment maybe on how should solve it. Please review the code again if there's anything else I should do as well.

**Review by frankgiordano — APPROVED** [review]
Looks good. Lets merge this and we can tweak it more after if need be!

**Comment by Nike682631** [comment]
> Looks good. Lets merge this and we can tweak it more after if need be!

We can ask leonid to approve as well or if you think it's good enough feel free to click on merge @frankgiordano

**Comment by frankgiordano** [comment]
i have merge this. Work on adding test sample programs.. thanks.

**Pull request merged**